Bottom: 2837bf264b71a773a3db1b22e2145e95d929c340
Top:    c864067a3cecdf526299c315e4fdd105ebe5ff2d
Author: Alexandre Oliva <aoliva@redhat.com>
Date:   2015-10-27 04:32:29 -0200

Test special member functions

---

diff --git a/libcc1/libcp1plugin.cc b/libcc1/libcp1plugin.cc
index db423e4..d53da61 100644
--- a/libcc1/libcp1plugin.cc
+++ b/libcc1/libcp1plugin.cc
@@ -863,7 +863,17 @@ plugin_start_new_class_type (cc1_plugin::connection *self,
 				      ctx->get_source_location (filename,
 								line_number));
 
-  return convert_out (ctx->preserve (type));
+  gcc_type class_type = convert_out (ctx->preserve (type));
+
+  if (strcmp (name, "lxtest") == 0) {
+    extern void lxtest_members (cc1_plugin::connection *self,
+				gcc_type class_type,
+				const char *filename,
+				unsigned int line_number);
+    lxtest_members (self, class_type, filename, line_number);
+  }
+
+  return class_type;
 }
 
 gcc_type
@@ -1255,6 +1265,35 @@ plugin_error (cc1_plugin::connection *,
   return convert_out (error_mark_node);
 }
 
+void
+lxtest_members (cc1_plugin::connection *self,
+		gcc_type class_type,
+		const char *filename,
+		unsigned int line_number) {
+  gcc_type void_type = plugin_void_type (self);
+  gcc_type int_type = plugin_int_type (self, 0, sizeof (int));
+  gcc_type_array argi = { 1, &int_type };
+  gcc_type fi2v = plugin_build_function_type (self, void_type, &argi, 0);
+  gcc_type mi2v = plugin_build_method_type (self, class_type, fi2v,
+					    gcc_cp_qualifiers (0),
+					    GCC_CP_REF_QUAL_NONE);
+  plugin_new_decl (self, "pL",
+		   gcc_cp_symbol_kind (GCC_CP_SYMBOL_FUNCTION |
+				       GCC_CP_FLAG_SPECIAL_FUNCTION),
+		   mi2v, 0, 2,
+		   filename, line_number);
+  plugin_new_decl (self, "C1",
+		   gcc_cp_symbol_kind (GCC_CP_SYMBOL_FUNCTION |
+				       GCC_CP_FLAG_SPECIAL_FUNCTION),
+		   mi2v, 0, 4,
+		   filename, line_number);
+  plugin_new_decl (self, "C2",
+		   gcc_cp_symbol_kind (GCC_CP_SYMBOL_FUNCTION |
+				       GCC_CP_FLAG_SPECIAL_FUNCTION),
+		   mi2v, 0, 6,
+		   filename, line_number);
+}
+
 
 
 // Perform GC marking.
